exist-core/src/main/java/org/exist/xquery/DynamicNameCheck.java

```diff
@@ -77,56 +80,52 @@ public Sequence eval(Sequence contextSequence, Item contextItem) throws XPathExc
                         ": required type is " + Type.getTypeName(test.getType()) +
                         "; got: " + Type.getTypeName(item.getType()) + ": " + item.getStringValue());
                 }
                 final Node node = ((NodeValue) item).getNode();
-                if (!test.matchesName(node))
-                    {throw new XPathException(expression, "Type error in expression: " +
-                        "required node name is " + getPrefixedNodeName(test.getName()) +
-                        "; got: " + getPrefixedNodeName(node));}
+                if (!test.matchesName(node)) {
+                    throw new XPathException(expression, ErrorCodes.XPTY0004, "Type error in expression: " +
+                        "required node name is " + getPrefixedNodeName(test) +
+                        "; got: " + getPrefixedNodeName((INode) node));
                 }
-                if (context.getProfiler().isEnabled()) {
-                    context.getProfiler().end(this, "", seq);
-                }
-                return seq;
-            } catch(final IllegalArgumentException iae) {
-                throw new XPathException(expression, iae);
             }
+            if (context.getProfiler().isEnabled()) {
+                context.getProfiler().end(this, "", seq);
+            }
+            return seq;
+        } catch(final IllegalArgumentException iae) {
+            throw new XPathException(expression, iae);
+        }
     }
 
-    private String getPrefixedNodeName(Node node) {
-        final String prefix = node.getPrefix();
-        if (prefix == null) {
-            final String nameSpace = node.getNamespaceURI();
-            if (nameSpace == null) {
-                return node.getNodeName();
-            } else {
-                return "{'" + nameSpace + "'}:" + node.getNodeName();
+    private String getPrefixedNodeName(final INode iNode) {
+        if (iNode instanceof Document) {
+            final Element documentElement = ((Document) iNode).getDocumentElement();
+            if (documentElement != null) {
+                return getPrefixedNodeName(true, ((INode) documentElement).getQName());
             }
-        } else if (prefix.isEmpty()) {
-            return "{''}:" + node.getNodeName();
-        } else {
-            return prefix + ":" + node.getLocalName();
         }
+        return getPrefixedNodeName(false, iNode.getQName());
     }
 
-    // TODO should be moved to QName
-    private String getPrefixedNodeName(QName name) {
-        final String prefix = name.getPrefix();
-        final String localName = name.getLocalPart();
-        if (prefix == null) {
-            final String namespaceURI = name.getNamespaceURI();
-            if (namespaceURI == null) {
-                return localName;
-            } else {
-                return "{'" + namespaceURI + "'}:" + localName;
-            }
-        } else if (prefix.isEmpty()) {
-            return "{''}:" + localName;
```

Comment:
i Remember that `{ .. }` is asserted in the XQTS suite?

Reply:
@dizzzz This is only used for the error message. I don't think that is checked in the tests... or if you can show me an example/link please?